Autopilot/PathManager/Inc/waypointManager.hpp

```diff
@@ -0,0 +1,175 @@
+/**
+ * Waypoint Manager Methods and Helpers
+ * Author: Dhruv Rawat
+ * Created: November 2020
+ * Last Updaetd: November 2020
+ */
+
+#ifndef WAYPOINT_MANAGER_H
+#define WAYPOINT_MANAGER_H
+
+#include <stdlib.h>
+
+#define ORBIT_FOLLOWING 0
+#define LINE_FOLLOWING 1
+
+#define PI 3.14159265359
+#define MAX_PATH_APPROACH_ANGLE PI/2
+
+#define PATH_BUFFER_SIZE 100
+
+//Waypoint Types
+#define DEFAULT_WAYPOINT 0 // Used to navigate flight path
+#define WAYPOINT_UNUSED 1
+#define HOLD_WAYPOINT 2 // Circle
+
+// Placeholder: 
+typedef struct {
+    /*
+        Contains:
+            long 
+            lat
+            alt
+            heading
+    */    
+} _Gps_Data;
+
+// Stores error codes for the waypoint manager
+enum _WaypointStatus {WAYPOINT_SUCCESS = 0, WAYPOINT_NOT_FOUND, WAYPOINT_PARAMETERS_NOT_DEFIND, UNDEFINED_FALIURE};
+
+// Used in the waypointBufferStatus array to signal which elements are free
+enum _WaypointBufferStatus {FREE = 0, FULL};
+
+//Used to specify the modification type when updating the waypointBuffer array
+enum _WaypointBufferUpdateType {ADD_WAYPOINT = 0, UPDATE_WAYPOINT, INSERT_WAYPOINT, DELETE_WAYPOINT};
+
+
+/*
+* Structure stores information about the waypoints along our path to the destination and back. 
+*/
+typedef struct {
+    struct _WaypointData * next;      // Next waypoint
+    struct _WaypointData * previous;  // Previous waypoint
+    long double latitude;             // Latitude of waypoint
+    long double longitude;            // Longitude of waypoint
+    int altitude;                     // Altitude of waypoint
+    float turnRadius;                 // If turn is commanded, this is the radius of the turn
+    char type;
+    char waypointId;                  // Id of the waypoint
```

Comment:
what's this ? How does the user know which ID to use ? How does it differ from the IDs mentioned in updatePathNodes() ?

Reply:
The waypoint IDs are used to have a reliable way of finding the waypoint objects in the waypointBuffer array. 

My understanding is the user or state machine will create the waypoint objects and pass them to the waypoint manager to update the waypointBuffer array. As a result, the state machine should be aware of the IDs of the different waypoints. 

The IDs in the struct and those in the updatePathNodes() methods are the same.